### DISCUSSION TITLE
Access request properties from within a server action.

### DISCUSSION THREAD

**Discussion by ken-spencer** [vercel/next.js#57949] [Ideas]
### Goals

It would be extremely handy if there was a way to access request properties  from within a server action. In my case for audit logging.  

### Non-Goals

1.
2.
3.

### Background

I am currently having to append the data I need to custom headers in middleware, but feels like a hack.

### Proposal

import request from "next/request"

...

let {geo, ip, nextUrl} = request();

**Comment by grichards** [comment]
Here's some code from my audit system that mostly gets around the lack of NextRequest in Server Actions:



```
// This code gets called from a useEffect one time when the site loads. I pass in the referrer.
async function createUser(referrer?: string) {
  const headersList = headers()
  const cookieStore = cookies()
  const userAgentStructure = {headers: headersList}
  const {isBot, browser, device, engine, os, cpu} = userAgent(userAgentStructure)
  let user

  try {
    user = await prisma.user.create({
      data: {
        ipAddress: headersList.get('x-real-ip') || headersList.get('x-forwarded-for'),
        referrer: referrer,
        userAgent: headersList.get('user-agent'),
        isBot: isBot,
        browserName: browser.name,
        browserVersion: browser.version,
        deviceModel: device.model,
        deviceType: device.type,
        deviceVendor: device.vendor,
        engineName: engine.name,
        engineVersion: engine.version,
        osName: os.name,
        osVersion: os.version,
        cpuArchitecture: cpu.architecture,
        headers: JSON.stringify(headersList, null, 2),
        cookies: JSON.stringify(cookieStore.getAll(), null, 2),
      },
    })
    cookieStore.set('etid', user.id as unknown as string)
  } catch (error) {
    const errorMessage = (error as Error).message
    sendMail({ subject: 'Error with createUser()', bodyHtml: errorMessage })
  }

  return user
}
```

**Reply by Nayanchandrakar** [reply]
thanks alot it help me

**Comment by Nayanchandrakar** [comment]
"use server";

import { serverUrl } from "@/lib/env-export";
import { cookies, headers } from "next/headers";
import { loginSchema, loginSchemaType } from "@/schema/zod-schema";
import { userAgent } from "next/server";

export const loginAction = async (formData: loginSchemaType) => {
  try {
    const validateFields = loginSchema?.safeParse(formData);

    if (!validateFields?.success) {
      return {
        error: "Invalid credentials",
      };
    }

    const headersList = headers();

    const userAgentStructure = { headers: headersList };
    const agent = userAgent(userAgentStructure);
    console.log(agent);

    const res = await fetch(serverUrl("/auth/login"), {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(formData),
    });

    const resData = await res.json();

    if (!res.ok) {
      return {
        error: resData?.error,
      };
    }

    const { data, success } = resData;
    cookies()?.set("token", data?.jwtToken);

    return {
      success,
    };
  } catch (error) {
    console.log(`[ERROR_FROM_LOGIN_ACTION]`, error);
    return {
      error: "Internal server error",
    };
  }
};
